### ISSUE TITLE
use of model.eval() and torch.no_grad()

### ISSUE THREAD

**Issue by jis3613** [mr-eggplant/SAR#2]
Hi, first of all congratulations on your great work and thanks for providing the code public!

I have couple questions about the usage of `model.eval()` and `with torch.no_grad()`

In main.py, for 'TENT' and 'EATA', you call the `validate` function defined on line 34, but for 'SAR' , the validation code is written inside the if state of  'SAR' from line 294 - 322. The difference is that for SAR, `model.eval()` and `with torch.no_grad()` are not used while both of them are used in the `validate` function.  

1. Is there particular reason why SAR does not need `model.eval()` and `with torch.no_grad()` ?
2. Would the batch normalization parameters be optimized for 'TENT' and 'EATA' with `model.eval()` and `torch.no_grad()` ?

Thank you!

**Comment by mr-eggplant** [comment]
Hi, thanks for your appreciation!

A1. I think there should be no difference for SAR, regardless of whether [_.eval()_ and _.no_grad()_] are used or not. Perhaps you could verify this?

A2. The norm parameters will be updated, as we use _@torch.enable_grad()_ in all of the following: [_tent.py, eata.py, sar.py_].

Actually, it seems that the existence of _.eval()_ and _.no_grad()_ is simply due to copying the validate function from another project :)

Hope that helps!

**Comment by jis3613** [comment]
Yeah I run the train code of SAR with `model.eval()` and `torch.no_grad()` inserted between line 300 and 301 of main.py for ResNet50-GN under mix_shifts corruption level 5 experiment setting and got 38.2 which is the same as the result reported in the paper. 
So I guess the use of `model.eval()` and `torch.no_grad()` does not really matter.
May be the line 294 - 322 of main.py can be replaced by just calling the `validate` function.

Thank you!

**Comment by mr-eggplant** [comment]
No problem at all:)
